Revise StorageClass docs (#42551)
kubernetes/website: content/en/docs/concepts/storage/storage-classes.md

with [volumes](/docs/concepts/storage/volumes/) and
[persistent volumes](/docs/concepts/storage/persistent-volumes) is suggested.

A StorageClass provides a way for administrators to describe the _classes_ of
storage they offer. Different classes might map to quality-of-service levels,
or to backup policies, or to arbitrary policies determined by the cluster
administrators. Kubernetes itself is unopinionated about what classes
represent.

The Kubernetes concept of a storage class is similar to “profiles” in some other
storage system designs.

<!-- body -->

## StorageClass objects

Each StorageClass contains the fields `provisioner`, `parameters`, and
`reclaimPolicy`, which are used when a PersistentVolume belonging to the
class needs to be dynamically provisioned to satisfy a PersistentVolumeClaim (PVC).

The name of a StorageClass object is significant, and is how users can
request a particular class. Administrators set the name and other parameters
of a class when first creating StorageClass objects.

> As an administrator, you can specify a default StorageClass that applies to any PVCs that

Author: I think the PVC page is more canonical TBH. The default StorageClass is a default that applies specifically to _PVCs_ without a StorageClass - right?

don't request any particular class to bind to; see the
[PersistentVolumeClaim concept](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims)
for more details.

Here's an example of a StorageClass:

{{% code_sample file="storage/storageclass-low-latency.yaml" %}}

## Default StorageClass

You can mark a StorageClass as the default for your cluster.
For instructions on setting the default StorageClass, see
[Change the default StorageClass](/docs/tasks/administer-cluster/change-default-storage-class/).

> When a PVC does not specify a `storageClassName`, the default StorageClass is

Reviewer: ```suggestion
When a PVC doesn't specify a `storageClassName`, the default StorageClass is
```

nit: use contractions like you did earlier

used.

If you set the
[`storageclass.kubernetes.io/is-default-class`](/docs/reference/labels-annotations-taints/#ingressclass-kubernetes-io-is-default-class)

> annotation to true on more than one StorageClass in your cluster, and you then

Reviewer: ```suggestion
annotation to `true` on more than one StorageClass in your cluster, and you then
```

Reviewer: Because it's the value in a manifest field I'd suggest the code formatting or bold, something to set it apart as the value

But this isn't a blocker!

Reviewer 2: Yes, code blocks help identify the value.

create a PersistentVolumeClaim with no `storageClassName` set, Kubernetes
picks which of the “default” StorageClasses to use based on which was created
more recently.

{{< note >}}
You should try to only have one StorageClass in your cluster that is
marked as the default. The reason that Kubernetes allows you to have
two (or more) default StorageClasses is to allow for seamless migration.
{{< /note >}}

You can create a PersistentVolumeClaim without specifying a `storageClassName`

> for the new PVC, and you can do so even when no default StorageClass exists

Reviewer 3: ```suggestion
, and you can do so even when no default StorageClass exists
```

Author: I prefer the original - what's not right with it?

in your cluster. In this case, the new PVC creates as you defined it, and the

> `storageClassName` of that PVC remains unset until default becomes available.

Reviewer 3: ```suggestion
`storageClassName` of that PVC remains unset until a default becomes available.
```

> You can have a cluster without any default StorageClass. If you don't mark any

Author: Basically, I was worried people might not get it. If we're sure nearly all readers will work it out with that paragraph omitted, let's omit it.

What do we think?

StorageClass as default (and your cloud provider, etc has not set one for you),
then Kubernetes cannot apply that defaulting for PersistentVolumeClaims that need
it.

If or when a default StorageClass becomes available, the control plane identifies any
existing PVCs without `storageClassName`. For the PVCs that either have an empty
value for `storageClassName` or do not have this key, the control plane then
updates those PVCs to set `storageClassName` to match the new default StorageClass.

> If you have an existing PVC where the `storageClassName` is `""`, and you configure

Reviewer 3: This statement is inconsistent with the previous one.
If `storageClassName: ""` is specified, a new default StorageClass is then added, what will happen?

Author: (this text is copied from https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retroactive-default-storageclass-assignment; if that text is wrong, we should fix both occurrences).

Reviewer 4: To me, who knows how default storage classes work, the PR text looks correct.

However, it would be great to distinguish between "empty value" (the filed is missing or `nil`) and empty string `""` more explicitly. How it's done in other API fields where `nil` and `""` means something different?

or to backup policies, or to arbitrary policies determined by the cluster
administrators. Kubernetes itself is unopinionated about what classes
represent.

The Kubernetes concept of a storage class is similar to “profiles” in some other
storage system designs.

<!-- body -->

## StorageClass objects

Each StorageClass contains the fields `provisioner`, `parameters`, and
`reclaimPolicy`, which are used when a PersistentVolume belonging to the
class needs to be dynamically provisioned to satisfy a PersistentVolumeClaim (PVC).

The name of a StorageClass object is significant, and is how users can
request a particular class. Administrators set the name and other parameters
of a class when first creating StorageClass objects, and the objects cannot
be updated once they are created.

As an administrator, you can specify a default StorageClass that applies to any PVCs that
don't request any particular class to bind to; see the
[PersistentVolumeClaim concept](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims)
for more details.

Here's an example of a StorageClass:

{{% code file="storage/storageclass-low-latency.yaml" %}}

> ## Default StorageClass

Author: The defaulting behavior is now generally available, so I thought it was worth mentioning here.

existing PVCs without `storageClassName`. For the PVCs that either have an empty
value for `storageClassName` or do not have this key, the control plane then
updates those PVCs to set `storageClassName` to match the new default StorageClass.
If you have an existing PVC where the `storageClassName` is `""`, and you configure
a default StorageClass, then this PVC will not get updated.

In order to keep binding to PVs with `storageClassName` set to `""`
(while a default StorageClass is present), you need to set the `storageClassName`
of the associated PVC to `""`.

This behavior helps administrators change the default StorageClass by removing the
old one first and then creating or setting another one. This brief window while
there is no default causes PVCs without `storageClassName` created at that time
to not have any default, and then the new default is applied shortly afterwards.

For instructions on setting the default StorageClass, see
[Change the default StorageClass](/docs/tasks/administer-cluster/change-default-storage-class/).
Note that certain cloud providers may already define a default StorageClass.

> ## Provisioner

Author: The table in this section is now in alphabetical order.

PersistentVolumes that are created manually and managed via a StorageClass will have
whatever reclaim policy they were assigned at creation.

## Volume expansion {#allow-volume-expansion}

PersistentVolumes can be configured to be expandable. This allows you to resize the
volume by editing the corresponding PVC object, requesting a new larger amount of
storage.

The following types of volumes support volume expansion, when the underlying
StorageClass has the field `allowVolumeExpansion` set to true.

{{< table caption = "Table of Volume types and the version of Kubernetes they require"  >}}

| Volume type          | Required Kubernetes version for volume expansion |
| :------------------- | :----------------------------------------------- |
| Azure File           | 1.11                                             |
| CSI                  | 1.24                                             |
| FlexVolume           | 1.13                                             |
| Portworx             | 1.11                                             |

> | rbd                  | 1.11                                             |

Author: This table is now in alphabetical order.

PersistentVolumes that are created manually and managed via a StorageClass will have
whatever reclaim policy they were assigned at creation.

## Volume Expansion {#allow-volume-expansion}

PersistentVolumes can be configured to be expandable. This allows you to resize the
volume by editing the corresponding PVC object, requesting a new larger amount of
storage.

The following types of volumes support volume expansion, when the underlying
StorageClass has the field `allowVolumeExpansion` set to true.

{{< table caption = "Table of Volume types and the version of Kubernetes they require"  >}}

| Volume type          | Required Kubernetes version for volume expansion |
| :------------------- | :----------------------------------------------- |
| Azure File           | 1.11                                             |
| CSI                  | 1.24                                             |
| FlexVolume           | 1.13                                             |
| GCE Persistent Disk  | 1.11                                             |

> | Portworx             | 1.11                                             |

Reviewer 3: `rbd` is dropped?

Author: Oops.
I re-added it.

StorageClass has the field `allowVolumeExpansion` set to true.

{{< table caption = "Table of Volume types and the version of Kubernetes they require"  >}}

| Volume type          | Required Kubernetes version for volume expansion |
| :------------------- | :----------------------------------------------- |
| Azure File           | 1.11                                             |
| CSI                  | 1.24                                             |
| FlexVolume           | 1.13                                             |
| GCE Persistent Disk  | 1.11                                             |
| Portworx             | 1.11                                             |
| rbd                  | 1.11                                             |

{{< /table >}}

{{< note >}}
You can only use the volume expansion feature to grow a Volume, not to shrink it.
{{< /note >}}

## Mount Options

PersistentVolumes that are dynamically created by a StorageClass will have the
mount options specified in the `mountOptions` field of the class.

If the volume plugin does not support mount options but mount options are

> specified, provisioning will fail. Mount options are i**not** validated on either

Reviewer 4: extra `i`?

Author: (fixed)

PersistentVolumes that are created manually and managed via a StorageClass will have
whatever reclaim policy they were assigned at creation.

## Volume Expansion {#allow-volume-expansion}

PersistentVolumes can be configured to be expandable. This allows you to resize the
volume by editing the corresponding PVC object, requesting a new larger amount of
storage.

The following types of volumes support volume expansion, when the underlying
StorageClass has the field `allowVolumeExpansion` set to true.

{{< table caption = "Table of Volume types and the version of Kubernetes they require"  >}}

> | Volume type          | Required Kubernetes version for volume expansion |

Author: This table is now in alphabetical order.

The following plugins support `WaitForFirstConsumer` with dynamic provisioning:

> - CSI volumes, provided that the specific CSI driver supports this

Reviewer 5: This list is woefully out of date. Maybe best to remove GCEPersistentDisk and just mention CSI (and Local below)

Author: Ideally, we can get a current list of in-tree providers that do provide that support. If it's in-tree, we try to document it _somewhere_. Even if it's also deprecated.

Reviewer 5: Off the top of my head, we have local, gce-pd, aws-ebs, azure-disk, azure-file. Maybe vsphere and ceph? @jsafrane can you confirm?

Author: The `awsElasticBlockStore` and `azureDisk` volume types are not included in Kubernetes v1.28 - is that right?

Reviewer 4: In 1.29, all cloud-based volume plugins are migrated to CSI, so just `local` remains in-tree and supports `WaitForFirstConsumer`. So CSI and `local` is more appropriate.

Author: Bear in mind: these are the docs for v1.28

Author: IMO it'd be nice to get this page updated ahead of the v1.29 release.

Author: (v1.29 is now released, and I've rebased)

Author: Is there any feature gate that lets you re-enable in tree support for azure file? I think we may have some work to do, as a project, to make sure the docs are up to date around those removals.

## Parameters

StorageClasses have parameters that describe volumes belonging to the storage
class. Different parameters may be accepted depending on the `provisioner`.
When a parameter is omitted, some default is used.

There can be at most 512 parameters defined for a StorageClass.
The total length of the parameters object including its keys and values cannot
exceed 256 KiB.

### AWS EBS

<!-- maintenance note: OK to remove all mention of awsElasticBlockStore once the v1.27 release of
Kubernetes has gone out of support -->

> Kubernetes {{< skew currentVersion >}} does not include a `awsElasticBlockStore` volume type.

Reviewer 4: This is IMO not accurate. Kubernetes indeed does not have AWS EBS volume plugin and we should strongly recommend users to use the CSI driver, however, we still support the old volume plugin API. A StorageClass with `provisioner: kubernetes.io/aws-ebs` still works, however, all the storage work is silently routed to the CSI driver. 

This applies to all cloud-based volume plugins in 1.29.

Author: OK. What's the accurate wording to tell users?

If they need to set up an add-on for `awsElasticBlockStore` to appear to work, perhaps we should signpost them to a page about that (that advice is yet to be written), or to https://kubernetes.io/docs/concepts/storage/volumes/#awselasticblockstore?

Reviewer 4: Yes, some wording in https://kubernetes.io/docs/concepts/storage/volumes/#awselasticblockstore would be appropriate. We should somehow say:

* The AWS EBS CSI driver must be installed for PVs with `awsElasticBlockStore` and StorageClasses with `provisioner: kubernetes.io/aws-ebs` to work. Even with the CSI driver.
* Kubernetes keeps this API only for backward compatibility to keep old PVs and StorageClasses working.
* We strongly suggest using the CSI driver directly in any new deployments.

Author: Sounds like a separate PR - would you or another contributor from SIG Storage be willing to file an issue to that effect @jsafrane ?
